{
 "metadata": {
  "name": "",
  "signature": "sha256:73117ad56155bc5bc1fd98a94515e4e917fd9d1b1e9962c554af681a1f02c358"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Pipelines: Putting it all together\n",
      "\n",
      "In this final lesson, we're going to put a lot of what we've learned together, literally, to create a **pipeline**.  \n",
      "\n",
      "## What is a pipeline?\n",
      "A pipeline is a sequence of analysis steps where the output of one step is used as the input for the next step.  The word 'pipeline' comes from an analogy with plumbing: in a pipe, fluid travels in only one direction.  In an analysis pipeline, the steps are executed in one direction, namely, from the beginning of the pipeline to the end.  You've probably come across pipelines in your everyday life, but just didn't call them that.  For example, you may have used a *getting ready for the workshop* pipeline this morning:\n",
      "\n",
      "<img src=\"pipeline1.png\" height=\"100\">\n",
      "\n",
      "In this pipeline, we first wake up, then eat breakfast, then go to the workshop.  The steps are followed in a specific, logical order.\n",
      "\n",
      "In bioinformatics, we often have to perform the same set of analysis steps on different datasets.  The tune sort of goes like this: get the data, preprocess the data, process the data, and visualize the data.  If we were to manually complete each one of these steps for each of the datasets it would be 1) very tedious 2) error-prone.  Instead, if we had a script to run all of these steps together it would be 1) very easy to analyze the data and 2) less error-prone.  Scripts that perform the steps of a pipeline are called **wrapper scripts** because they \"wrap\" all the steps together into one script.  Analyzing data with wrapper scripts makes the analysis reproducible.  This is because running the script multiple times on the same input should always produce the same output.\n",
      "\n",
      "<img src=\"meme.jpg\" width=\"500\">\n",
      "\n",
      "\n",
      "## Writing your first wrapper script\n",
      "We're going to write a wrapper script in Python to analyze cd-hit data.  The pipeline will have the following steps:\n",
      "\n",
      "1. Run `cd-hit` on a protein sequences file.\n",
      "2. Run a `cd-hit` tool called `clstr_size_stat.pl` to calculate the distribution of cluster sizes\n",
      "3. Run a Python script to plot the distribution of cluster sizes\n",
      "\n",
      "Thus, the input to our pipeline will be a file containing protein sequences and the output will be a figure in pdf format.  Graphically, it looks like this:\n",
      "\n",
      "<img src=\"pipeline2.png\" height=\"700\">\n",
      "\n",
      "We're going to shove all of these steps into a single script called `cd_hit_wrapper.py`.\n",
      "\n",
      "Before we start writing any code, it's time for a little philosophy about writing code.  When writing code, there are two main approaches: [top-down and bottom-up](http://en.wikipedia.org/wiki/Top-down_and_bottom-up_design).  In top-down programming, you first outline the major steps you need to do, then you successively divide the steps into smaller and smaller steps until you are left with the simplest of steps that you can easily code.  Bottom-up programming is the opposite: you code all of the basic steps, and then combine them together to make the more complex steps.  We're going to use the top-down approach because it has the advantage that we'll be able to see the big picture of our program early on. This means that during the development process, we may be calling scripts that do not yet exist.  By the end of the process, we will have coded all the steps so the program will be functional.\n",
      "\n",
      "### Step 1\n",
      "Start with this wrapper script:\n",
      "\n",
      "       $ wget http://genomic.wustl.edu/bioinformatics/python/lessons/cd_hit_wrapper.py\n",
      "   \n",
      "   \n",
      "  Follow the directions inside the script to learn how the pipeline works and write the code.  You will add code wherever there is a `TODO` comment.\n",
      "  \n",
      "  You'll need the `proteins.faa` file from before.  Copy it to your current working directory or download it again:\n",
      "  \n",
      "      $ wget http://genomic.wustl.edu/bioinformatics/proteins.faa\n",
      "\n",
      "**Hints:**\n",
      "* Use the Python function **`subprocess.call`** to call an external command from within a Python script.  The syntax is:\n",
      "\n",
      "        subprocess.call(list_of_arguments)\n",
      "    \n",
      "  where `list_of_arguments` is a Python list of the words in the command.\n",
      "    For example, if you wanted to execute\n",
      "    \n",
      "        $ cd-hit -i proteins.faa -o clustered_proteins.faa\n",
      "        \n",
      "    `list_of_arguments` would be\n",
      "    \n",
      "        [\"cd-hit\", \"-i\", \"proteins.faa\", \"-o\", \"clustered_proteins.faa\"]\n",
      "        \n",
      "    and the `subprocess.call` code would be\n",
      "    \n",
      "        subprocess.call([\"cd-hit\", \"-i\", \"proteins.faa\", \"-o\", \"clustered_proteins.faa\"])\n",
      "        \n",
      "* When you want to capture the stdout of an external command, the syntax for `subprocess.call` is a little different:\n",
      "\n",
      "        subprocess.call(list_of_arguments, stdout=file_object)\n",
      "\n",
      "  where `file_object` is a Python file object for a file that you opened for writing.  Refer to the previous [reading and writing files lesson](http://nbviewer.ipython.org/github/mollygibson/bioinformatics_wkshp_washu/blob/master/python/lessons/Python7_FileIO.ipynb) for help with file objects.\n",
      "\n",
      "### Step 2\n",
      "After you've finished writing `cd_hit_wrapper.py`, we need to write `plot_cluster_size.py` since the wrapper script called this script and it's not currently written.  Start with this script:\n",
      "\n",
      "        $ wget http://genomic.wustl.edu/bioinformatics/python/lessons/plot_cluster_size.py\n",
      "\n",
      "Follow the directions inside the script to learn how it works and write the code.  You will add code wherever there is a `TODO` comment.\n",
      "\n",
      "**Hint:**\n",
      "* Refer to the previous [plotting lesson](http://nbviewer.ipython.org/github/mollygibson/bioinformatics_wkshp_washu/blob/master/python/lessons/Python9_Matplotlib.ipynb) for help with plotting.\n",
      "\n",
      "### Step 3\n",
      "When you're finished with the previous two steps, run your pipeline:\n",
      "\n",
      "        $ python3 cd_hit_wrapper.py proteins.fa 0.8\n",
      "        \n",
      "  Use Cyberduck to transfer the pdf file it created from the remote computer to your computer and open the file.  Try running the pipeline with different sequence identity thresholds.\n",
      "  \n",
      "<img src=\"stopgo.jpeg\" width=\"100\">"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}